zone/task_manager.cpp

```diff
@@ -1639,18 +1553,262 @@ bool TaskManager::LoadClientState(Client *client, ClientTaskState *client_task_s
+			client_task_state->m_active_task.accepted_time,
+			client_task_state->m_active_task.updated
+		);
+
+		TaskInformation *p_task_data = task_manager->m_task_data[client_task_state->m_active_task.task_id];
+		if (p_task_data != nullptr) {
+			for (int i = 0; i < p_task_data->activity_count; i++) {
+				if (client_task_state->m_active_task.activity[i].activity_id >= 0) {
+					LogTasksDetail(
+						"[LoadClientState] -- character_id [{}] task [{}] activity_id [{}] done_count [{}] activity_state [{}] updated [{}] sequence [{}]",
+						character_id,
+						client_task_state->m_active_task.task_id,
+						client_task_state->m_active_task.activity[i].activity_id,
+						client_task_state->m_active_task.activity[i].done_count,
+						client_task_state->m_active_task.activity[i].activity_state,
+						client_task_state->m_active_task.activity[i].updated,
+						p_task_data->sequence_mode
+					);
+				}
+			}
+		}
 	}
 
-	// TODO: shared
+	// shared task
+	LogTasksDetail(
+		"[LoadClientState] m_active_shared_task task_id is [{}] slot [{}]",
+		client_task_state->m_active_shared_task.task_id,
+		client_task_state->m_active_shared_task.slot
+	);
+	if (client_task_state->m_active_shared_task.task_id != TASKSLOTEMPTY) {
+		client_task_state->UnlockActivities(character_id, client_task_state->m_active_shared_task);
+	}
+
+	// quests (max 20 or 40 depending on client)
 	for (auto &active_quest : client_task_state->m_active_quests) {
 		if (active_quest.task_id != TASKSLOTEMPTY) {
 			client_task_state->UnlockActivities(character_id, active_quest);
 		}
 	}
 
-	LogTasks("[LoadClientState] for Character ID [{}] DONE!", character_id);
+	LogTasksDetail("[LoadClientState] for Character ID [{}] DONE!", character_id);
+	LogTasksDetail("---", character_id);
 
 	return true;
 }
+
+void TaskManager::SyncClientSharedTaskState(Client *c, ClientTaskState *cts)
+{
+	LogTasksDetail(
+		"[SyncClientSharedTaskState] Syncing client shared task state"
+	);
+
+	SyncClientSharedTaskWithPersistedState(c, cts);
+	SyncClientSharedTaskRemoveLocalIfNotExists(c, cts);
+}
+
+void TaskManager::SyncClientSharedTaskWithPersistedState(Client *c, ClientTaskState *cts)
+{
+	auto character_tasks = CharacterTasksRepository::GetWhere(
+		database,
+		fmt::format("charid = {} ORDER BY acceptedtime", c->CharacterID())
+	);
+
+	for (auto &character_task: character_tasks) {
+		if (character_task.type == TASK_TYPE_SHARED) {
+			auto st = SharedTaskMembersRepository::GetWhere(
+				database,
+				fmt::format(
+					"character_id = {}",
+					c->CharacterID()
+				)
+			);
+
+			if (!st.empty()) {
+				int64 shared_task_id = st[0].shared_task_id;
+				auto  activities     = SharedTaskActivityStateRepository::GetWhere(
+					database,
+					fmt::format(
+						"shared_task_id = {}",
+						shared_task_id
+					)
+				);
+
+				ClientTaskInformation *shared_task = nullptr;
+				shared_task = &cts->m_active_shared_task;
+
+				// has active shared task
+				if (shared_task) {
```

Comment:
@Akkadius Should this be `cts->HasActiveSharedTask()`? Since `shared_task` is just a pointer to the array that will always be true instead of checking for a valid task_id

Reply:
Yeah good call. That's why I had created HasActiveSharedTask to begin with; good eye